can't compile with --config=apple_debug

### What happened?

```
% bazel build --config=apple_debug //solvers/...
```
fails with 
```
ERROR: /private/var/tmp/_bazel_russt/9c100c86c1cc4c543d01c45f7a63d5c6/external/nlopt_internal/BUILD.bazel:197:11: Compiling src/algs/stogo/stogo.cc failed: undeclared inclusion(s) in rule '@nlopt_internal//:stogo':
this rule is missing dependency declarations for the following files included by 'src/algs/stogo/stogo.cc':
  'external/nlopt_internal/src/algs/stogo/global.h'
  'external/nlopt_internal/src/algs/stogo/tools.h'
  'external/nlopt_internal/src/algs/stogo/linalg.h'
...
ERROR: /private/var/tmp/_bazel_russt/9c100c86c1cc4c543d01c45f7a63d5c6/external/nlopt_internal/BUILD.bazel:246:11: Compiling src/algs/ags/ags.cc failed: undeclared inclusion(s) in rule '@nlopt_internal//:ags':
this rule is missing dependency declarations for the following files included by 'src/algs/ags/ags.cc':
  'external/nlopt_internal/src/algs/ags/solver.hpp'
  'external/nlopt_internal/src/algs/ags/data_types.hpp'
  'external/nlopt_internal/src/algs/ags/evolvent.hpp'
  'external/nlopt_internal/src/algs/ags/local_optimizer.hpp'
...
```

```
bazel build --config=apple_debug //multibody/...
```
throws those errors and also a bunch from conex
```
bazel-out/darwin_arm64-dbg/bin/external/conex_internal/conex/drake_vendor/conex/triangular_matrix_workspace.cc:6:11: error: unknown type name 'TriangularMatrixWorkspace'; did you mean '::conex::TriangularMatrixWorkspace'?
using T = TriangularMatrixWorkspace;
          ^~~~~~~~~~~~~~~~~~~~~~~~~
          ::conex::TriangularMatrixWorkspace
external/conex_internal/conex/triangular_matrix_workspace.h:14:8: note: '::conex::TriangularMatrixWorkspace' declared here
struct TriangularMatrixWorkspace {
       ^
bazel-out/darwin_arm64-dbg/bin/external/conex_internal/conex/drake_vendor/conex/triangular_matrix_workspace.cc:38:1: error: use of undeclared identifier 'TriangularMatrixWorkspace'; did you mean '::conex::TriangularMatrixWorkspace'?
TriangularMatrixWorkspace::TriangularMatrixWorkspace(
^~~~~~~~~~~~~~~~~~~~~~~~~
::conex::TriangularMatrixWorkspace
external/conex_internal/conex/triangular_matrix_workspace.h:14:8: note: '::conex::TriangularMatrixWorkspace' declared here
struct TriangularMatrixWorkspace {
       ^
bazel-out/darwin_arm64-dbg/bin/external/conex_internal/conex/drake_vendor/conex/triangular_matrix_workspace.cc:38:28: error: cannot define or redeclare 'TriangularMatrixWorkspace' here because namespace 'conex' does not enclose namespace 'TriangularMatrixWorkspace'
TriangularMatrixWorkspace::TriangularMatrixWorkspace(
~~~~~~~~~~~~~~~~~~~~~~~~~~~^
bazel-out/darwin_arm64-dbg/bin/external/conex_internal/conex/drake_vendor/conex/triangular_matrix_workspace.cc:40:22: error: use of undeclared identifier 'supernode_size_'
    : supernode_size(supernode_size_) {
                     ^
bazel-out/darwin_arm64-dbg/bin/external/conex_internal/conex/drake_vendor/conex/triangular_matrix_workspace.cc:41:3: error: use of undeclared identifier 'N'
  N = std::accumulate(supernode_size.begin(), supernode_size.end(), 0);
  ^
bazel-out/darwin_arm64-dbg/bin/external/conex_internal/conex/drake_vendor/conex/triangular_matrix_workspace.cc:41:23: error: use of undeclared identifier 'supernode_size'
  N = std::accumulate(supernode_size.begin(), supernode_size.end(), 0);
                      ^
bazel-out/darwin_arm64-dbg/bin/external/conex_internal/conex/drake_vendor/conex/triangular_matrix_workspace.cc:41:47: error: use of undeclared identifier 'supernode_size'
  N = std::accumulate(supernode_size.begin(), supernode_size.end(), 0);
                                              ^
bazel-out/darwin_arm64-dbg/bin/external/conex_internal/conex/drake_vendor/conex/triangular_matrix_workspace.cc:42:3: error: use of undeclared identifier 'variable_to_supernode_'
  variable_to_supernode_.resize(N);
  ^
bazel-out/darwin_arm64-dbg/bin/external/conex_internal/conex/drake_vendor/conex/triangular_matrix_workspace.cc:42:33: error: use of undeclared identifier 'N'
  variable_to_supernode_.resize(N);
                                ^
bazel-out/darwin_arm64-dbg/bin/external/conex_internal/conex/drake_vendor/conex/triangular_matrix_workspace.cc:43:3: error: use of undeclared identifier 'variable_to_supernode_position_'
  variable_to_supernode_position_.resize(N);
  ^
bazel-out/darwin_arm64-dbg/bin/external/conex_internal/conex/drake_vendor/conex/triangular_matrix_workspace.cc:43:42: error: use of undeclared identifier 'N'
  variable_to_supernode_position_.resize(N);
                                         ^
bazel-out/darwin_arm64-dbg/bin/external/conex_internal/conex/drake_vendor/conex/triangular_matrix_workspace.cc:55:3: error: use of undeclared identifier 'snodes'
  snodes.resize(path_.size());
  ^
bazel-out/darwin_arm64-dbg/bin/external/conex_internal/conex/drake_vendor/conex/triangular_matrix_workspace.cc:55:17: error: use of undeclared identifier 'path_'
  snodes.resize(path_.size());
                ^
bazel-out/darwin_arm64-dbg/bin/external/conex_internal/conex/drake_vendor/conex/triangular_matrix_workspace.cc:56:19: error: use of undeclared identifier 'snodes'
  for (auto& si : snodes) {
                  ^
bazel-out/darwin_arm64-dbg/bin/external/conex_internal/conex/drake_vendor/conex/triangular_matrix_workspace.cc:57:15: error: use of undeclared identifier 'supernode_size'
    si.resize(supernode_size.at(cnt));
              ^
bazel-out/darwin_arm64-dbg/bin/external/conex_internal/conex/drake_vendor/conex/triangular_matrix_workspace.cc:58:25: error: use of undeclared identifier 'supernode_size'
    for (int i = 0; i < supernode_size.at(cnt); i++) {
                        ^
bazel-out/darwin_arm64-dbg/bin/external/conex_internal/conex/drake_vendor/conex/triangular_matrix_workspace.cc:59:18: error: use of undeclared identifier 'N'
      if (var >= N) {
                 ^
bazel-out/darwin_arm64-dbg/bin/external/conex_internal/conex/drake_vendor/conex/triangular_matrix_workspace.cc:62:18: error: use of undeclared identifier 'path_'
      si.at(i) = path_.at(cnt).at(i);
                 ^
bazel-out/darwin_arm64-dbg/bin/external/conex_internal/conex/drake_vendor/conex/triangular_matrix_workspace.cc:63:7: error: use of undeclared identifier 'variable_to_supernode_'
      variable_to_supernode_[var] = cnt;
      ^
fatal error: too many errors emitted, stopping now [-ferror-limit=]
```
cc @frankpermenter 

### Version

master

### What operating system are you using?

macOS 14 (Sonoma)

### What installation option are you using?

compiled from source code using Bazel

### Relevant log output

```shell
russt@TRI-GQ16T2RQ2R drake % brew config
HOMEBREW_VERSION: 4.2.9
ORIGIN: https://github.com/Homebrew/brew
HEAD: e5fefd73cd97cd36ae3af29551f529ae59b333d6
Last commit: 7 days ago
Core tap JSON: 25 Feb 21:15 UTC
Core cask tap JSON: 25 Feb 21:15 UTC
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_EDITOR: emacs
HOMEBREW_MAKE_JOBS: 12
Homebrew Ruby: 3.1.4 => /opt/homebrew/Library/Homebrew/vendor/portable-ruby/3.1.4/bin/ruby
CPU: dodeca-core 64-bit arm_blizzard_avalanche
Clang: 15.0.0 build 1500
Git: 2.39.3 => /Applications/Xcode.app/Contents/Developer/usr/bin/git
Curl: 8.4.0 => /usr/bin/curl
macOS: 14.3.1-arm64
CLT: 15.1.0.0.1.1700200546
Xcode: 15.2
Rosetta 2: false
```

--- a/doc/_pages/bazel.md
+++ b/doc/_pages/bazel.md
@@ -132,21 +132,6 @@ For a Python unittest that uses ``drake_py_unittest``, for example:
 bazel test bindings/pydrake:py/symbolic_test --test_output=streamed --nocache_test_results --test_arg=--trace=user --test_arg=TestSymbolicVariable
 ```
 
-## Debugging on macOS
-
-On macOS, DWARF debug symbols are emitted to a ``.dSYM`` file.  The Bazel
-``cc_binary`` and ``cc_test`` rules do not natively generate or expose this
-file, so we have implemented a workaround in Drake, ``--config=apple_debug``.
-This config turns off sandboxing, which allows a ``genrule`` to access the
-``.o`` files and process them into a ``.dSYM``.  Use as follows:
-
-```
-bazel build --config=apple_debug path/to/my:binary_or_test_dsym
-lldb ./bazel-bin/path/to/my/binary_or_test
-```
-
-For more information, see [https://github.com/bazelbuild/bazel/issues/2537](https://github.com/bazelbuild/bazel/issues/2537).
-
 # Updating BUILD files
 
 Please use the "``buildifier``" tool to format edits to ``BUILD`` files (in the

--- a/doc/_pages/profiling.md
+++ b/doc/_pages/profiling.md
@@ -80,25 +80,3 @@ iterations it runs on each benchmark, which will skew the performance metrics
 you see in the profiler. [You may want to manually specify the number of
 iterations to run on each
 benchmark](https://stackoverflow.com/questions/61843343/how-to-special-case-the-number-of-iterations-in-google-benchmark).
-
-# Debugging and profiling on macOS
-
-On macOS, DWARF debug symbols are emitted to a ``.dSYM`` file.  The Bazel
-``cc_binary`` and ``cc_test`` rules do not natively generate or expose this
-file, so we have implemented a workaround in Drake, ``--config=apple_debug``.
-This config turns off sandboxing, which allows a ``genrule`` to access the
-``.o`` files and process them into a ``.dSYM``.  Use as follows:
-
-```
-bazel build --config=apple_debug path/to/my:binary_or_test_dsym
-lldb ./bazel-bin/path/to/my/binary_or_test
-```
-
-Profiling on macOS can be done by building with the debug symbols and then running
-```
-xcrun xctrace record -t "Time Profiler" --launch ./bazel-bin/path/to/my/binary_or_test
-```
-This will generate a `.trace` file that can be opened in the Instruments app:
-```
-open -a Instruments myfile.trace
-```

--- a/tools/cc_toolchain/BUILD.bazel
+++ b/tools/cc_toolchain/BUILD.bazel
@@ -8,12 +8,6 @@ config_setting(
 )
 
 config_setting(
-    name = "apple_debug",
-    constraint_values = ["@platforms//os:osx"],
-    values = {"compilation_mode": "dbg"},
-)
-
-config_setting(
     name = "debug",
     values = {"compilation_mode": "dbg"},
 )

--- a/tools/cc_toolchain/bazel.rc
+++ b/tools/cc_toolchain/bazel.rc
@@ -12,16 +12,3 @@ build --action_env=CCACHE_DISABLE=1
 # When compiilng Drake as an external package, this rcfile is not loaded and we
 # won't promote warnings to errors by default.
 build --define=DRAKE_WERROR=ON
-
-### Debug symbols on OS X. ###
-# See https://github.com/bazelbuild/bazel/issues/2537
-# and https://github.com/bazelbuild/bazel/issues/14294
-build:apple_debug --spawn_strategy=standalone
-build:apple_debug --genrule_strategy=standalone
-build:apple_debug --compilation_mode=dbg
-build:apple_debug --notrim_test_configuration
-
-# We don't actually use APPLE_DEBUG in code. It's just here to invalidate any
-# sandboxed .o files that might be in cache if a developer tried to build
-# with --compilation_mode=dbg instead of --config=apple_debug.
-build:apple_debug --copt="-DAPPLE_DEBUG"

--- a/tools/skylark/drake_cc.bzl
+++ b/tools/skylark/drake_cc.bzl
@@ -128,24 +128,6 @@ def _platform_copts(rule_copts, rule_gcc_copts, rule_clang_copts, cc_test = 0):
         ],
     })
 
-def _dsym_command(name):
-    """Returns the command to produce .dSYM on macOS, or a no-op on Linux."""
-    return select({
-        "//tools/cc_toolchain:apple_debug": (
-            "dsymutil -f $(location :" + name + ") -o $@ 2> /dev/null"
-        ),
-        "//conditions:default": (
-            "touch $@"
-        ),
-    })
-
-def _dsym_srcs(name):
-    """Returns the input for making a .dSYM on macOS, or a no-op on Linux."""
-    return select({
-        "//tools/cc_toolchain:apple_debug": [":" + name],
-        "//conditions:default": [],
-    })
-
 def _check_library_deps_blacklist(name, deps):
     """Report an error if a library should not use something from deps."""
     if not deps:
@@ -792,19 +774,6 @@ def drake_cc_binary(
         **kwargs
     )
 
-    # Also generate the OS X debug symbol file for this binary.
-    tags = kwargs.pop("tags", [])
-    native.genrule(
-        name = name + "_dsym",
-        srcs = _dsym_srcs(name),
-        outs = [name + ".dSYM"],
-        output_to_bindir = 1,
-        testonly = testonly,
-        tags = tags + ["dsym"],
-        visibility = ["//visibility:private"],
-        cmd = _dsym_command(name),
-    )
-
     if "@gtest//:main" in deps:
         fail("Use drake_cc_googletest to declare %s as a test" % name)
 
@@ -883,18 +852,6 @@ def drake_cc_test(
             "-no_deduplicate",
         ],
         **kwargs
-    )
-
-    # Also generate the OS X debug symbol file for this test.
-    native.genrule(
-        name = name + "_dsym",
-        srcs = _dsym_srcs(name),
-        outs = [name + ".dSYM"],
-        output_to_bindir = 1,
-        testonly = kwargs["testonly"],
-        tags = ["dsym"],
-        visibility = ["//visibility:private"],
-        cmd = _dsym_command(name),
     )
 
 def drake_cc_googletest(

